### PR TITLE
feat: Add options for Transition Altitude

### LINKED ISSUE #2068: Load Transition Altitude from DB
**Mod Version**

master of 23rd of Nov 20

**Describe the bug**

Transition ALT on the takeoff or Approach page is defaulted to 10000ft. It should be taken from the nav database and inserted automatically depending on the DEP/ARR airport.

**References**

IRL experience

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### LINKED ISSUE #2572: [REQUEST] Default Transition Alt/Lvl
**Is your feature request related to a problem? Please describe.**
No

**Describe the feature you'd like to see implemented**
Flying only in the USA you have to put in transition altitude into performance takeoff and Approach phase pages. Would be nice to implement an optional trans altitude in A320X options. so one can avoid this.  Currently it is defaulted to 10,000.

**References**

**Additional context**

Discord username (if different from GitHub):

### LINKED ISSUE #2813: [REQUEST] Transition altitude preference in MCDU otions
**Is your feature request related to a problem? Please describe.**
Not a problem

**Describe the feature you'd like to see implemented**
In the same menu as fuel units please add a preference for transition altitude for those countries / regions where it is the same everywhere 
Have this preferred transition altitude populate the PERF pages for Takeoff and approach (i.e will always populate FL180 for North America)

**References**
Not applicable 

**Additional context**
Not applicable 

Discord username (if different from GitHub): nnumber

### PR DESCRIPTION
**Pull request by Kimbyeoungjang**
Fixes #2572
Fixes #2813
Fixes #2068

## Summary of Changes

As a title, Add MCDU options for Transition Altitude
I'll also add auto calculate based on location(to use airport ICAO code)

- [x] Add Options for Trans ALT
- [x] Build Transition Altitude logic
- [x] Build Auto Transition Altitude(based on ICAO code and custom database) logic
- [x] Connect to Flybywire API
- [x] Add transLVL logic
- [x] Update when spawn on runway (not C&D)
- [ ] fix minor problem(change my code to reworked code)

## Screenshots (if necessary)

image
image

## References

image
ICAO countries

https://mediawiki.ivao.aero/index.php?title=How_to_determine_Transition_level
Transition Altitude & Level Calculate

## Additional context

Discord username (if different from GitHub):

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by Kimbyeoungjang** [comment]
Ignore upper commits, It just update XD

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js:29]:
```suggestion
            ["DEP TRANS ALT", `{cyan}[${storedDepartTransAlt}]*{end}`],
            [""],
            ["ARR TRANS ALT", `{cyan}[${storedArrivalTransAlt}]*{end}`],

```

Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js:29]: So we are consistent through our pages

**Review by MisterChocker — CHANGES_REQUESTED** [review]

**Comment by MisterChocker** [comment]
What do you think about making the cyan inactive text small font?

**Comment by Kimbyeoungjang** [comment]
> What do you think about making the cyan inactive text small font?

that sounds good! i'll do that :)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js:34]:
```suggestion
        mcdu.setTemplate([
            ["A32NX OPTIONS REALISM"],
            ["", "", "TRANS ALT MODE"],
            [Auto],
            [""],
            [Manual],
            ["", "", "DEFAULT MANUAL TA"],
            [""],
            ["TA DEPATURE"],
            [`{cyan}*[${storedDepartTransAlt}]{end}`],
            ["TA ARRIVAL"],
            [`{cyan}*[${storedArrivalTransAlt}]{end}`],
            [""],
            ["<RETURN"]
        ]);

```

Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js:34]: would need to adjust the buttons below.

**Review by MisterChocker — CHANGES_REQUESTED** [review]

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js
Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js:34]: maybe make manual font small if on auto mode.

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_TransitionManager.js
Comment (MisterChocker) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_TransitionManager.js:5]:
```suggestion
        if (mode !== "AUTO") {
```

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_TransitionManager.js
Comment (MisterChocker) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_TransitionManager.js:14]:
```suggestion
        if (mode === "AUTO") {
```

**Review by MisterChocker — COMMENTED** [review]
I know you are still working on it, but I noticed these two small things. In js we want to these `===` comparison over `==`, the reason for that is performance because `===` only compares the values itself and not the types as `==` does.

I'd also suggest using integers for the types, like 0 for auto and 1 for manual.
Would look like this then:
`const mode = parseInt(NXDataStore.get("CONFIG_TRANSALT", "0"));`

**Comment by Kimbyeoungjang** [comment]
@MisterChocker Ok!

And can i ask about nx datastore? 
I’m not sure I do a right job. Because this is first time to use nx data store.

**Comment by MisterChocker** [comment]
Yes looking about right, just keep in mind you want read and write to it via strings.

**Comment by Kimbyeoungjang** [comment]
> Yes looking about right, just keep in mind you want read and write to it via strings.

Thx mate! Here is 01:00AM XD have a good night~

thank you for your help :)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js:58]:
```suggestion
                mcdu.addNewMessage(NXSystemMessages.notAllowed);
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_TRANSITION.js:70]:
```suggestion
                mcdu.addNewMessage(NXSystemMessages.notAllowed);
```

**Review by MisterChocker — COMMENTED** [review]
To make messages compatible with new system

**Comment by ErikNL80** [comment]
I see a spelling error: depature should be departure

**Comment by Kimbyeoungjang** [comment]
> I see a spelling error: depature should be departure

Thx i fix that :)

**Comment by Kimbyeoungjang** [comment]
So, First, TransLVL logic is inserted but it is disabled. because some country have fixed transLVL. So just disable at this time, and i'll open other PR.

QA Testers! when you test this, please wait more than 3 min. this code working every 3 mins because I worried about API server. So, just test as a normal flight, but please wait 3 min. just programming MCDU or climb slow.

And, Please test on Cold&Dark(Route is not inserted) and spawn on Runway(Route inserted)

**Comment by Kimbyeoungjang** [comment]
Also, check the MCDU realism option. there is a manual & auto option for transAlt

File: .github/CHANGELOG.md
Comment (NathanInnes) [inline .github/CHANGELOG.md:27]: Can you shorten this to 1 line please?

**Comment by St54Kevin** [comment]
Doesnt this also fix #2068 ?

**Comment by MisterChocker** [comment]
> Doesnt this also fix #2068 ?

yes it should.

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_TransitionManager.js
Comment (MisterChocker) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_TransitionManager.js:26]: why do you need to check (have/use an update routine?)
What about calling the update/change when changing departure/arrival?

Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/A32NX_Core/A32NX_TransitionManager.js:26]: sure! i fix it today :)

**Comment by Kimbyeoungjang** [comment]
Wait until takeoff perf enhancement merged to master.

**Comment by Kimbyeoungjang** [comment]
my work is done. this is my best :(...

this is QA Tier 2

QA Testers, Please test like below

First Test, manually insert flightplan(from cold&dark) and check the mcdu performance page - takeoff & approach phase, check that there is a trans ALT inserted after you insert flight plan

and flying to destination with checking PFD, QNH section. when you climb, pass transition altitude, PFD QNH should flashing
and last, when you approach, pass transition altitude, PFD STD should flashing.

Second Test, start from runway or gate. But you should insert flightplan automatically via worldmap
before you check mcdu performance page, you should enter init page first.

if you test first, you only check transALT is inserted on Performance page.

and last, for developers, if you see this, please provide some good idea because i am idiot.
I can't get out in initpage code to A32NX Core folder because mcdu.flightPlanManager is not working on A32NX Core folder.

**Comment by Kimbyeoungjang** [comment]
change draft due to merge mcdu performance page rework

**Comment by Kimbyeoungjang** [comment]
just close this PR, and using reworked logic. will open other pr.

**Pull request closed without merging**
